Develop: build and install directories + configure script

hrabbel opened the pull request:
- simplified build and install using configure script
- build directory can be chosen by argument to configure script: -DBUILDDIR
- install directory can be chosen by argument to configure script or cmake: -DINSTALLDIR_LEMONADE 
- libraries are build at /path/to/build/directory/lib/  and installed into /path/to/install/directory/lib/LeMonADE 
- binaries are build at /path/to/build/directory/bin/  and installed into /path/to/install/directory/bin/
- headers are installed into /path/to/install/directory/include/LeMonADE/

Bondoki committed:
* changed the output path in cmake
* providing a simple script for installing 'configure'
* introducing an optional CMAKE-variable INSTALLDIR_LEMONADE
* updated the README

hrabbel committed:
Edit the configure script as well as CMakeLists.txt files for easier build and install

*The wrapper Makefile was removed. Instead it is now created by the configure script
*The configure script can take two optional arguments: One to set the build directory
 and one to set the install directory. This way the steps of building and installing
 are more clearly distinguished.
*The install directory is used only when 'make install' is issued. The install directory
 can be set with the option -DINSTALLDIR_LEMONADE=/path/to/install. It defaults to /usr/local .
 The library is then installed into /path/to/install/lib/LeMonADE/, the headers
 into /path/to/install/include/LeMonADE/, the binaries into /path/to/install/bin/.
*The build directory can be set with -DBUILDDIR. It can be any directory, also far away from
 the source.
*Details on build and install procedure in README.md

hrabbel committed:
Updated the README.md to match the changes after last commit.

hrabbel committed:
Merge branch 'bondoki-develop' into develop.

This provides a simplified build and installation procedure.

Bondoki commented on configure:
```diff
@@ -0,0 +1,82 @@
+clean:
+	@- (cd ${BUILDDIR} >/dev/null 2>&1 && cmake .. >/dev/null 2>&1)
+	@- \$(MAKE) --silent -C ${BUILDDIR} clean || true
+	@- \$(RM) ./${BUILDDIR}/Makefile
```
Remove line 71-75

Bondoki commented on configure:
```diff
@@ -0,0 +1,82 @@
+	@ (cd ${BUILDDIR} >/dev/null 2>&1 && cmake ${INSTALLDIR_ARGUMENT} ${PWD_PATH})
+
+clean:
+	@- (cd ${BUILDDIR} >/dev/null 2>&1 && cmake .. >/dev/null 2>&1)
```
replace: @- (cd ${BUILDDIR} >/dev/null 2>&1 && cmake .. >/dev/null 2>&1)
to: @- (cd ${BUILDDIR} >/dev/null 2>&1)

Bondoki replied:
or remove

Bondoki commented on configure:
```diff
@@ -0,0 +1,82 @@
+	@- \$(RM) ./${BUILDDIR}/projects
+	@- \$(RM) ./${BUILDDIR}/CMake*
+	@- \$(RM) ./${BUILDDIR}/*.cmake
+
```
add: .PHONY: clean

Bondoki commented on configure:
```diff
@@ -0,0 +1,82 @@
+
+install:
+	@ \$(MAKE) -C ${BUILDDIR} install 
+
```
add: .PHONY: install

Bondoki commented on projects/SimpleSimulator/CMakeLists.txt:
```diff
@@ -1,8 +1,10 @@
 
 target_link_libraries(SimpleSimulator LeMonADE)
+
+install(TARGETS SimpleSimulator DESTINATION bin)
```
remove this line!

Bondoki commented on configure:
```diff
@@ -0,0 +1,82 @@
+            usage         
+            exit
+            ;;
+            
```
also provide option -DLEMONADE_TESTS=ON/OFF

Bondoki commented on src/CMakeLists.txt:
```diff
@@ -37,6 +37,6 @@ IF (project_build_static)
 	#TARGET_LINK_LIBRARIES(staticlib ${project_LIBS})
 	SET_TARGET_PROPERTIES(libLeMonADE-static PROPERTIES VERSION "${APPLICATION_VERSION_MAJOR}.${APPLICATION_VERSION_MINOR}" OUTPUT_NAME ${project_BIN} CLEAN_DIRECT_OUTPUT 1)
-	INSTALL(TARGETS libLeMonADE-static DESTINATION lib)
+	INSTALL(TARGETS libLeMonADE-static DESTINATION lib/LeMonADE)
```
instead of "lib/LeMonADE" only "lib"

Bondoki commented on projects/Examples/example4/CMakeLists.txt:
```diff
@@ -2,7 +2,7 @@ cmake_minimum_required (VERSION 2.6)
 
 
```
replace by:
if (NOT DEFINED LEMONADE_INCLUDE_DIR)
    message("LEMONADE_INCLUDE_DIR is not provided. Build might fail. Use -DLEMONADE_INCLUDE_DIR=/path/to/installed/include/header/LeMonADE/ or install to default location")
endif()

if (NOT DEFINED LEMONADE_LIBRARY_DIR)
    message("LEMONADE_LIBRARY_DIR is not provided. Build might fail. Use -DLEMONADE_LIBRARY_DIR=/path/to/installed/lib/LeMonADE/ or install to default location")
endif()

include_directories (${LEMONADE_INCLUDE_DIR})
link_directories (${LEMONADE_LIBRARY_DIR})

Bondoki commented on CMakeLists.txt:
```diff
@@ -48,39 +48,47 @@ SET (CMAKE_VERBOSE_MAKEFILE 0) # Use 1 for debugging, 0 for release
 #
-#MESSAGE(${PROJECT_SOURCE_DIR})
-#SET (MAINFOLDER ${PROJECT_SOURCE_DIR})
 SET (LEMONADE_DIR ${PROJECT_SOURCE_DIR})
```
add:
SET (LEMONADE_INCLUDE_DIR ${LEMONADE_DIR})
SET (LEMONADE_LIBRARY_DIR "${CMAKE_BINARY_DIR}/lib")

hrabbel committed:
Fix errors and improve configure script and cmake files

* configure script now has option for enabling tests
* made targets in wrapper makefile (created by configure script) PHONY, because otherwise it does not work on all platforms
* defined extra variables LEMONADE_INCLUDE_DIR and LEMONADE_LIBRARY_DIR in CMakeLists.txt files.
  This is useful when compiling parts of the projects in different directories, because it allows for specifying the
  location of headers and lemonade library to link against.
* In the installation process, the loki lib headers are now also installed, such that after installation all required
  headers are now available on the system.

hrabbel commented:
Ok, I added the suggestions in the last commit.

Bondoki commented on README.md:
```diff
@@ -1,39 +1,81 @@
+  /given/installation/path/lib/LeMonADE/, or if no path was specified to
+  /usr/local/lib/LeMonADE/. Similar for the header files, which go to
+  /given/installation/path/include/LeMonADE/ or /usr/local/include/LeMonADE/.
+  The binaries are installed into /given/installation/path/bin/ or /usr/local/bin/.
```
Remove this line. We only install the libs and include

Bondoki commented on README.md:
````diff
@@ -1,39 +1,81 @@
     cd build
     cmake ..
     make docs
+```
````
add this lines too:

``` sh
     # generates the docs
     ./configure -DINSTALLDIR_LEMONADE=/path/to/install/LeMonADE/ -DBUILDDIR=/path/to/build/
     make docs
```

Bondoki commented on configure:
```diff
@@ -0,0 +1,91 @@
+clean:
+	@- \$(MAKE) --silent -C ${BUILDDIR} clean || true
+.PHONY: clean
+
```
Add this lines for the documentation:

``` sh
     docs: 
          @  \$(MAKE) -C ${BUILDDIR} docs
     .PHONY : docs
```

Bondoki commented on CMakeLists.txt:
```diff
@@ -48,39 +48,49 @@ SET (CMAKE_VERBOSE_MAKEFILE 0) # Use 1 for debugging, 0 for release
+INSTALL(DIRECTORY "${LEMONADE_DIR}/include/" DESTINATION  "include")
 
 #
 # Add Documentation Targets
```
replace the doxygen stuff by this lines:

``` sh
     SET (DOC_INPUT_FILE_PATH "${LEMONADE_DIR}/docs/")
     SET (DOC_OUTPUT_FILE_PATH "${INSTALLDIR_LEMONADE}/docs/")

    FIND_PACKAGE (Doxygen)
    IF (DOXYGEN_FOUND)
        MESSAGE("Build documentation with: make docs")
        IF (EXISTS ${DOC_INPUT_FILE_PATH})
            MESSAGE("Existing File documentation with doxygen")
            configure_file(${DOC_INPUT_FILE_PATH}doxygen.conf ${DOC_OUTPUT_FILE_PATH}doxygen.conf @ONLY)
            configure_file(${DOC_INPUT_FILE_PATH}mainpage.dox ${DOC_OUTPUT_FILE_PATH}mainpage.dox @ONLY)
            configure_file(${DOC_INPUT_FILE_PATH}figures/ProgramStructure.jpg         ${DOC_OUTPUT_FILE_PATH}figures/ProgramStructure.jpg @ONLY)
            ADD_CUSTOM_TARGET(
            docs
                ${DOXYGEN_EXECUTABLE} ${DOC_INPUT_FILE_PATH}doxygen.conf
            WORKING_DIRECTORY ${DOC_OUTPUT_FILE_PATH}
                COMMENT "Generating doxygen project documentation." VERBATIM
            )
        ELSE (EXISTS ${DOC_INPUT_FILE_PATH})
            ADD_CUSTOM_TARGET(docs COMMENT "Doxyfile not found. Please generate a doxygen configuration file to use this target." VERBATIM)
        ENDIF (EXISTS ${DOC_INPUT_FILE_PATH})
    ELSE (DOXYGEN_FOUND)
        ADD_CUSTOM_TARGET(docs COMMENT "Doxygen not found. Please install doxygen to use this target." VERBATIM)
    ENDIF (DOXYGEN_FOUND)
```

Bondoki commented:
Also update "INSTALL"

Bondoki commented:
Also update Doxygen.conf and main page.

hrabbel commented:
Closing this, because I am restructuring my repo. New pull request to come
